processor/logdedupprocessor/processor.go

```diff
@@ -89,21 +94,45 @@ func (p *logDedupProcessor) ConsumeLogs(_ context.Context, pl plog.Logs) error {
 		for j := 0; j < rl.ScopeLogs().Len(); j++ {
 			sl := rl.ScopeLogs().At(j)
 			scope := sl.Scope()
+			logs := sl.LogRecords()
 
-			for k := 0; k < sl.LogRecords().Len(); k++ {
-				logRecord := sl.LogRecords().At(k)
-				// Remove excluded fields if any
-				p.remover.RemoveFields(logRecord)
+			logs.RemoveIf(func(logRecord plog.LogRecord) bool {
+				return p.matchFunc(ctx, logRecord, scope, resource, sl, rl)
+			})
```

Comment:
The matchFunc signature ended up being very verbose. Maybe we should just bypass `RemoveIf` altogether when appropriate.

```go
if p.condition == nil {
	// Don't worry about RemoveIf
	p.remover.RemoveFields(logRecord)
	p.aggregator.Add(resource, scope, logRecord)
	continue
}

// now worry about RemoveIf
logs.RemoveIf(func(logRecord plog.LogRecord) bool {
	logCtx := ottllog.NewTransformContext(logRecord, scope, resource, sl, rl)
	logMatch, err := p.condition.Eval(ctx, logCtx)
	if err != nil {
		p.logger.Error("error matching condition", zap.Error(err))
		return false
	}
	if logMatch {
		p.remover.RemoveFields(logRecord)
		p.aggregator.Add(resource, scope, logRecord)
	}
	return logMatch
})
```

You might also consider moving the two common lines into a function to ensure we don't update one and not the other later.